db/blob/blob_source_test.cc

```diff
@@ -1020,6 +1021,279 @@ TEST_F(BlobSourceTest, MultiGetBlobsFromCache) {
+
+  std::unique_ptr<BlobFileCache> blob_file_cache(new BlobFileCache(
+      backing_cache.get(), &immutable_options, &file_options, column_family_id,
+      blob_file_read_hist, nullptr /*IOTracer*/));
+
+  BlobSource blob_source(&immutable_options, db_id_, db_session_id_,
+                         blob_file_cache.get());
+
+  ConcurrentCacheReservationManager* cache_res_mgr =
+      immutable_options.blob_cache->GetCacheReservationManager();
+  ASSERT_NE(cache_res_mgr, nullptr);
+
+  ReadOptions read_options;
+  read_options.verify_checksums = true;
+
+  std::vector<PinnableSlice> values(keys.size());
+
+  {
+    read_options.fill_cache = false;
+
+    for (size_t i = 0; i < num_blobs; ++i) {
+      ASSERT_OK(blob_source.GetBlob(
+          read_options, keys[i], blob_file_number, blob_offsets[i], file_size,
+          blob_sizes[i], kNoCompression, nullptr /* prefetch_buffer */,
+          &values[i], nullptr /* bytes_read */));
+      ASSERT_EQ(cache_res_mgr->GetTotalReservedCacheSize(), 0);
+      ASSERT_EQ(cache_res_mgr->GetTotalMemoryUsed(), 0);
+    }
+  }
+
+  {
+    read_options.fill_cache = true;
+
+    // num_blobs is 16, so the total blob cache usage is less than a single
+    // dummy entry. Therefore, cache reservation manager only reserves one dummy
+    // entry here.
+    uint64_t blob_bytes = 0;
+    for (size_t i = 0; i < num_blobs; ++i) {
+      ASSERT_OK(blob_source.GetBlob(
+          read_options, keys[i], blob_file_number, blob_offsets[i], file_size,
+          blob_sizes[i], kNoCompression, nullptr /* prefetch_buffer */,
+          &values[i], nullptr /* bytes_read */));
+      blob_bytes += blob_sizes[i];
+      ASSERT_EQ(cache_res_mgr->GetTotalReservedCacheSize(), kSizeDummyEntry);
+      ASSERT_EQ(cache_res_mgr->GetTotalMemoryUsed(), blob_bytes);
+      ASSERT_EQ(cache_res_mgr->GetTotalMemoryUsed(),
+                options_.blob_cache->GetUsage());
+    }
+  }
+
+  {
+    OffsetableCacheKey base_cache_key(db_id_, db_session_id_, blob_file_number,
+                                      file_size);
+    size_t blob_bytes = options_.blob_cache->GetUsage();
+
+    for (size_t i = 0; i < num_blobs; ++i) {
+      CacheKey cache_key = base_cache_key.WithOffset(blob_offsets[i]);
+      options_.blob_cache->Erase(cache_key.AsSlice());
+      if (i == num_blobs - 1) {
+        // The last blob is not in the cache. cache_res_mgr should not reserve
+        // any space for it.
+        ASSERT_EQ(cache_res_mgr->GetTotalReservedCacheSize(), 0);
+      } else {
+        ASSERT_EQ(cache_res_mgr->GetTotalReservedCacheSize(), kSizeDummyEntry);
+      }
+      ASSERT_EQ(cache_res_mgr->GetTotalMemoryUsed(),
+                blob_bytes - blob_sizes[i]);
+      ASSERT_EQ(cache_res_mgr->GetTotalMemoryUsed(),
+                options_.blob_cache->GetUsage());
+      blob_bytes -= blob_sizes[i];
+    }
+  }
+}
+
+TEST_F(BlobSourceCacheReservationTest, IncreaseCacheReservationOnFullCache) {
```

Comment:
nit: is there any way to shorten these two tests by refactoring their similar parts into some functions of BlobSourceCacheReservationTest? Just want to make each test focus on the essential part related to cache_res_mgr.

Reply:
Done.